apollo-router/src/plugins/rhai.rs

```diff
@@ -0,0 +1,507 @@
+        let service = service.map_request(move |request: RouterRequest| {
+            let mut headers = headers_cloned.lock().expect("headers lock poisoned");
+            *headers = request.context.request.headers().clone();
+
+            request
+        });
+
+        service
+            .map_response(move |mut response: RouterResponse| {
+                let extensions = block_on(async { response.context.extensions().await.clone() });
+                let (headers, context) =
+                    match this.run_rhai_script(FUNCTION_NAME, headers_map, extensions) {
+                        Ok(res) => res,
+                        Err(err) => {
+                            return plugin_utils::RouterResponse::builder()
+                                .errors(vec![Error::builder()
+                                    .message(format!("RHAI plugin error: {}", err))
+                                    .build()])
+                                .context(response.context)
+                                .build()
+                                .into();
+                        }
+                    };
+                *(block_on(async { response.context.extensions_mut().await })) = context;
+
+                for (header_name, header_value) in &headers {
+                    response
+                        .response
+                        .headers_mut()
+                        .append(header_name, header_value.clone());
+                }
+
+                response
+            })
+            .boxed()
+    }
+
+    fn query_planning_service(
+        &mut self,
+        service: BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError>,
+    ) -> BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError> {
+        const FUNCTION_NAME: &str = "query_planning_service";
+        if !self
+            .ast
+            .iter_fn_def()
+            .any(|fn_def| fn_def.name == FUNCTION_NAME)
+        {
+            tracing::debug!("RHAI plugin: no query_planning_service function found");
+            return service;
+        }
+
+        let this = self.clone();
+        let headers_map = Arc::new(Mutex::new(HeaderMap::new()));
+        let headers_cloned = headers_map.clone();
+        let service = service.map_request(move |request: QueryPlannerRequest| {
+            let mut headers = headers_cloned.lock().expect("headers lock poisoned");
+            *headers = request.context.request.headers().clone();
+
+            request
+        });
+
+        service
+            .map_response(move |response: QueryPlannerResponse| {
+                let extensions = block_on(async { response.context.extensions().await.clone() });
+                let (_headers, context) =
+                    match this.run_rhai_script(FUNCTION_NAME, headers_map, extensions) {
+                        Ok(res) => res,
+                        Err(err) => {
+                            // there is no way to return an error properly
```

Comment:
why is there no way to return an error properky?

Reply:
It's just I'm not very happy with the way we raise an error here. Basically in a `QueryPlannerResponse` there is not http_response or equivalent. Only a `QueryPlan` and a context. Then I have to pass the error in the context and catch it later